components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1243 @@
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use crate::codec;
+use crate::codec::mysql::{self, Tz};
+use crate::codec::TEN_POW;
+use crate::codec::{Error, Result};
+use crate::expr::{EvalConfig, EvalContext, Flag, SqlMode};
+
+use tidb_query_datatype::FieldTypeTp;
+
+use bitfield::bitfield;
+use chrono::prelude::*;
+
+const MIN_TIMESTAMP: i64 = 0;
+const MAX_TIMESTAMP: i64 = (1 << 31) - 1;
+const MICRO_WITDH: usize = 6;
+
+fn bytes_to_u32(input: &[u8]) -> Result<u32> {
```

Comment:
this function is unsafe

Reply:
parse expect a `&str`, in rust, `&str` should contain valid utf8 char seq, which means that `parse` may assume that its param is valid utf8 str. so if we pass a invalid char seq to it, its result may not valid.

Reply:
Done